src/Raven.Client/Documents/Session/AsyncDocumentSession.Query.cs

```diff
@@ -9,8 +9,13 @@ public partial class AsyncDocumentSession
     {
         public IRavenQueryable<T> Query<T, TIndexCreator>() where TIndexCreator : AbstractCommonApiForIndexes, new()
         {
-            var indexCreator = new TIndexCreator();
-            return Query<T>(indexCreator.IndexName, null, indexCreator.IsMapReduce);
+            if (DocumentSession.IndexMetadataCache.TryGetValue(typeof(TIndexCreator), out var index) == false)
+            {
+                index = new DocumentSession.IndexMetadataCacheItem(new TIndexCreator());
+                DocumentSession.IndexMetadataCache.TryAdd(typeof(TIndexCreator), index);
+            }
```

Comment:
This is repeated in many locations. Better to have a single method for that which does all the work.

Reply:
Done.